src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectSystem.cs

```diff
@@ -159,23 +160,18 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
 
         try
         {
-            var tasks = new List<Task>();
-
-            var projectsThatNeedRestore = new ConcurrentSet<string>();
-            foreach (var projectToLoad in projectPathsToLoadOrReload)
-            {
-                tasks.Add(Task.Run(async () =>
+            var projectsThatNeedRestore = await ProducerConsumer<string>.RunParallelAsync(
+                source: projectPathsToLoadOrReload,
+                produceItems: static async (projectToLoad, callback, args, cancellationToken) =>
                 {
-                    var projectNeedsRestore = await LoadOrReloadProjectAsync(projectToLoad, toastErrorReporter, buildHostProcessManager, cancellationToken);
+                    var projectNeedsRestore = await args.@this.LoadOrReloadProjectAsync(
+                        projectToLoad, args.toastErrorReporter, args.buildHostProcessManager, cancellationToken);
 
                     if (projectNeedsRestore)
-                    {
-                        projectsThatNeedRestore.Add(projectToLoad.Path);
-                    }
-                }, cancellationToken));
-            }
-
-            await Task.WhenAll(tasks);
+                        callback(projectToLoad.Path);
+                },
+                args: (@this: this, toastErrorReporter, buildHostProcessManager),
+                cancellationToken).ConfigureAwait(false);
```

Comment:
>cancellationToken).ConfigureAwait(false); [](http://example.com/codeflow?start=16&length=41)

Does the RestoreProjectsAsync call below convert the ImmutableArray to an ImmutableHashSet, and then convert it to a string[]?

Reply:
updated to passa long the array.